docs/decisions/0017-reimplement-asset-processing.rst

```diff
@@ -0,0 +1,268 @@
+
+     - ``paver update_assets --skip-collect``
+
+       A Python-defined task that calls out to each build stage.
+
+     - ``assets/build.sh``
+
+       A Bash script that contains all build stages, with subcommands available for running each stage separately. Its command-line interface inspired by Tutor's ``openedx-assets`` script. The script will be runnable on any POSIX system, including macOS and Ubuntu and it will linted for common shell scripting mistakes using `shellcheck <https://www.shellcheck.net>`_.
+     
+   * - + **Build stage 1: Copy npm-installed assets** from node_modules to other folders in edx-platform. They are used by certain especially-old legacy LMS & CMS frontends that are not set up to work with npm directly.
+
+     - ``paver update_assets --skip-collect``
+
+       Implemented in Python within update_assets. There is no standalone command for it.
+
+     - ``assets/build.sh npm``
+
+       Pure Bash reimplementation.
+ 
+   * - + **Build stage 2: Copy XModule fragments** from the xmodule source tree over to input directories for Webpack and SCSS compilation. This is required for a hard-coded list of old XModule-style XBlocks. This is not required for new pure XBlocks, which include (or pip-install) their assets into edx-platform as ready-to-serve JS/CSS/etc fragments.
+
+     - ``paver process_xmodule_assets``, or
+
+       ``xmodule_assets``
+
+       Equivalent paver task and console script, both pointing at to an application-level Python module. That module inspects attributes from legacy XModule-style XBlock classes in order to determine which static assets to copy and what to name them.
+
+     - ``assets/build.sh xmodule``
+
+       A Bash implementation of XModule asset copying. The aforementioned attributes will be moved from the XModule-style XBlock classes into a simple static JSON file, which the Bash script will be able to read.
+       
+       The initial implementation of build.sh may just point at ``xmodule_assets``.
+   
+   * - + **Build stage 3: Run Webpack** in order to to shim, minify, otherwise process, and bundle JS modules. This requires a call to the npm-installed ``webpack`` binary.
+
+     - ``paver webpack``
+
+       Python wrapper around a call to webpack. Invokes the ``./manage.py [lms|cms] print_setting`` multiple times in order to determine Django settings, adding which can add 20+ seconds to the build.
+
+     - ``assets/build.sh webpack``
+
+       Bash wrapper around a call to webpack. The script will accept parameters for Django settings rather than looking them up. Open edX distributions, such as Tutor, can choose how to supply the Django-setting-derived parameters in an efficient manner.
```

Comment:
Could the re-implementation include a single management command to dump a JSON file or such containing the current value of all the Django settings currently used to build static assets which will now need to be provided as parameters?  That would simplify the migration, and allow for wrapper scripts that could be directly dropped in for the old paver commands.

Reply:
I like that it would presumably hold a list of the exact settings we need for assets and wouldn't just print everything.  This would help to isolate which settings the asset build depends on.

Reply:
That looks good, thanks!

Reply:
Added.